tests/test_df_resampling.py

```diff
@@ -1,109 +1,167 @@
 
-    # Date range we want to end up with
-    index = pd.date_range('2015-01-01', '2015-12-31', freq='7D')
+    def test_weekly_to_monthly(self):
+        """Test weekly to monthly model time-step."""
+        input_df = make_df('W')
+        model_index = make_model_index('M')
 
-    df7D = align_and_resample_dataframe(df2D, index)
+        with pytest.raises(ResamplingError):
+            input_resampled = align_and_resample_dataframe(input_df, model_index)
 
-    assert df7D.index.freq == index.freq
-    assert df7D[0] == 1.5
-    assert df7D[1] == 5.0
-    assert df7D[2] == 8.5
+    def test_weekly_to_two_weekly(self):
+        """Test weekly to two-weekly model time-step."""
+        input_df = make_df('W')
+        model_index = make_model_index('2W')
 
-    assert df7D.index[0] == index[0]
-    assert df7D.index[-1] == index[-1]
+        input_resampled = align_and_resample_dataframe(input_df, model_index, resample_func='mean')
 
-    # Required range now one day after start of our data
-    index = pd.date_range('2015-01-02', '2015-12-31', freq='7D')
+        pd.testing.assert_index_equal(input_resampled.index, model_index)
+        pd.testing.assert_frame_equal(input_resampled, input_resampled.resample('2W').mean())
 
-    # This should raise an error because the resulting index
-    # won't align with the desired on.
-    with pytest.raises(ValueError):
-        df7D = align_and_resample_dataframe(df2D, index)
 
-    # Required range now two days after start of our data
-    index = pd.date_range('2015-01-03', '2015-12-31', freq='7D')
+class TestDownSampling:
+    """Test for down-sampling a dataframe to higher frequency model time-step."""
+    def test_monthly_to_daily(self):
+        """Test monthly data to daily model time-step."""
+        input_df = make_df('M')
+        model_index = make_model_index('D')
 
-    df7D = align_and_resample_dataframe(df2D, index)
+        input_resampled = align_and_resample_dataframe(input_df, model_index, resample_func='ffill')
 
-    assert df7D.index[0] == index[0]
-    # The end does not necessarily align
-    # assert df7D.index[-1] == index[-1]
+        pd.testing.assert_index_equal(input_resampled.index, model_index)
+        pd.testing.assert_frame_equal(input_resampled, input_resampled.resample('D').mean())
 
-    assert df7D.index.freq == index.freq
-    assert df7D[0] == 2.5
-    assert df7D[1] == 6.0
-    assert df7D[2] == 9.5
+    def test_weekly_to_daily(self):
+        """Test weekly aligned to daily model time-step."""
+        # Note that we have to choose a week starting on Wednesday because 2015-01-01 is a Wednesday.
+        input_df = make_df('W-WED')
+        model_index = make_model_index('D')
 
-    # Required range now one day before start of our data
-    index = pd.date_range('2014-12-31', '2015-12-31', freq='7D')
+        input_resampled = align_and_resample_dataframe(input_df, model_index, resample_func='ffill')
```

Comment:
Not sure if the resample_func kwarg has any impact for some of these test as it only seems to be used by `align_and_resample_dataframe` when the Dataframe index and model index are both types of `DateOffset`

Reply:
I've reworked the resampling stuff here. I changed the name of the keyword to reflect that is only used for down-sampling. I have then parametrised the tests that use down-sampling to test with "mean" and "max" to check this is working.

Reply:
Thanks, just gone through it again and it all looks good